File: internal/pool/pool.go
// Package pool reconciles the account-global rate_limits % across the per-session statusline
// snapshots — the one thing ccusage can't see. Each snapshot is frozen at a session's last API
// turn, so a window is reconciled across sessions: newest still-plausible reset, max used% on it.
package pool

import (
	"os"
	"path/filepath"

	"github.com/SeanLF/ccpool/internal/paths"
	"github.com/SeanLF/ccpool/internal/profile"
	"github.com/SeanLF/ccpool/internal/rb"
)

// Week is the 7-day weekly window length in seconds.
const Week = 7 * 86400

// Stale is how old a snapshot may be before the raw % is distrusted (CCPOOL_STALE_SECS, default
// 120s). During active use the statusline re-renders seconds apart, so a trustworthy snapshot is
// seconds old; past this the caller extrapolates or tiers down.
func Stale() int64 {
	if v, ok := os.LookupEnv("CCPOOL_STALE_SECS"); ok {
		return int64(rb.ToI(v))
	}
	return 120
}

// Window is a reconciled rate-limit window: the used % and its reset epoch.
type Window struct {
	Used  float64
	Reset int64
}

// LoadSnapshots reads every per-session snapshot (falling back to the bare cache file if no
// per-session files exist), dropping anything unreadable or non-object. Order matches Ruby's sorted
// Dir.glob so callers that take the first match agree.
func LoadSnapshots() []map[string]any {
	files, _ := filepath.Glob(paths.SnapshotGlob())
	if len(files) == 0 {
		if _, err := os.Stat(paths.SnapshotCache()); err == nil {
			files = []string{paths.SnapshotCache()}
		}
	}
	var out []map[string]any
	for _, f := range files {
		b, err := os.ReadFile(f)
		if err != nil {
			continue
		}
		if m := rb.ParseObject(b); m != nil {
			out = append(out, m)
		}
	}
	return out
}

// GetWindow reconciles one account-global window across snapshots frozen at differing staleness:
// the newest still-plausible reset, and the MAX used% on it (monotonic within a window -> freshest).
// Guards the leak bug (#52326: used% carrying a resets_at epoch) and clamps garbage. ok=false when
// no plausible window exists.
func GetWindow(snaps []map[string]any, key string, now, maxAhead int64) (Window, bool) {
	var lives []Window
	for _, d := range snaps {
		w := digNum(d, "rate_limits", key)
		if w == nil {
			continue
		}
		u, ok := rb.Num(w["used_percentage"])
		if !ok || u < 0 || u >= 10000 {
			continue
		}
		r, ok := rb.Num(w["resets_at"])
		if !ok {
			continue
		}
		reset := int64(r)
		if reset <= now || reset > now+maxAhead {
			continue
		}
		lives = append(lives, Window{Used: min(u, 100), Reset: reset})
	}
	if len(lives) == 0 {
		return Window{}, false
	}
	// Newest plausible reset, then the max used% on it (monotonic within a window -> freshest).
	// used% is clamped to [0,100] above and maxReset is one of lives', so a 0.0 seed is safe.
	maxReset := lives[0].Reset
	for _, l := range lives {
		maxReset = max(maxReset, l.Reset)
	}
	maxUsed := 0.0
	for _, l := range lives {
		if l.Reset == maxReset {
			maxUsed = max(maxUsed, l.Used)
		}
	}
	return Window{Used: maxUsed, Reset: maxReset}, true
}

// DataAge is the seconds since the freshest snapshot across all sessions. ok=false if none.
func DataAge(snaps []map[string]any, now int64) (int64, bool) {
	var newest int64
	found := false
	for _, d := range snaps {
		if c, ok := rb.Num(d["captured_at"]); ok {
			if ci := int64(c); !found || ci > newest {
				newest, found = ci, true
			}
		}
	}
	if !found {
		return 0, false
	}
	return now - newest, true
}

// Pace measures used% against the fraction of the window's ACTIVITY WEIGHT elapsed (Profile). With
// the default even profile that is the plain elapsed fraction. delta>0 = ahead of pace.
type Pace struct {
	ElapsedPct float64
	Delta      float64
	ToReset    int64
}

func GetPace(used float64, reset, now int64) Pace {
	pct := profile.Load().ElapsedFraction(reset-Week, now, reset) * 100
	return Pace{ElapsedPct: pct, Delta: used - pct, ToReset: reset - now}
}

// Weekly / FiveHour reconcile the two windows with their age attached (age via the caller).
func Weekly(snaps []map[string]any, now int64) (Window, bool) {
	return GetWindow(snaps, "seven_day", now, Week+86400)
}

func FiveHour(snaps []map[string]any, now int64) (Window, bool) {
	return GetWindow(snaps, "five_hour", now, 6*3600)
}

// --- helpers ---

func digNum(m map[string]any, keys ...string) map[string]any {
	cur := m
	for _, k := range keys {
		next, ok := cur[k].(map[string]any)
		if !ok {
			return nil
		}
		cur = next
	}
	return cur
}

File: pool.rb
# frozen_string_literal: true

# Core pool read: reconcile the account-global rate_limits % across the per-session
# statusline snapshots, and compute pace vs the fraction of the 7-day window elapsed.
# This is the one thing ccusage can't see (it never gets rate_limits) -- our floor.
require "json"

module Pool
  CACHE = File.expand_path(ENV["USAGE_CACHE"] || "~/.claude/usage-cache.json")
  GLOB  = CACHE.sub(/\.json\z/, "-*.json")
  WEEK  = 7 * 86_400
  # How old a snapshot may be before we distrust the raw % and fall to the estimate tier.
  # During active use the statusline re-renders the % multiple times/minute (event-driven,
  # ~300ms debounce), so a trustworthy snapshot is SECONDS old. 2min covers active use
  # comfortably; past it we extrapolate (via accrued ccusage cost) to catch a headless
  # fan-out burning unseen (it renders no statusline). Tight is safer here, not looser.
  STALE = (ENV["CCPOOL_STALE_SECS"] || "120").to_i

  module_function

  def load_snapshots
    files = Dir.glob(GLOB)
    files = [CACHE] if files.empty? && File.exist?(CACHE)
    files.filter_map { |f| (JSON.parse(File.read(f)) rescue nil) }.select { |d| d.is_a?(Hash) }
  end

  # Reconcile one account-global window across snapshots frozen at differing staleness:
  # newest still-plausible reset, MAX used% on it (monotonic within a window -> freshest).
  # Guards the leak bug (#52326: used% carrying a resets_at epoch) and clamps garbage.
  def window(snaps, key, now, max_ahead)
    live = snaps.filter_map do |d|
      w = d.dig("rate_limits", key)
      next unless w.is_a?(Hash) && (u = w["used_percentage"]).is_a?(Numeric) && u >= 0 && u < 10_000
      r = w["resets_at"]
      next unless r.is_a?(Numeric) && r > now && r <= now + max_ahead

      { used: [u.to_f, 100.0].min, reset: r }
    end
    return if live.empty?

    reset = live.map { _1[:reset] }.max
    { used: live.select { _1[:reset] == reset }.map { _1[:used] }.max, reset: reset }
  end

  # Seconds since the freshest snapshot across all sessions (nil if none).
  def data_age(snaps = load_snapshots, now = Time.now.to_i)
    cap = snaps.filter_map { |d| d["captured_at"] if d["captured_at"].is_a?(Numeric) }.max
    cap && (now - cap)
  end

  def fresh?(snaps, now)
    (age = data_age(snaps, now)) && age <= STALE
  end

  # Raw reconciled window + its age. nil only if NO plausible window exists at all;
  # staleness is the CALLER's decision (it drives the fresh/estimated/stale tiers).
  def weekly(now = Time.now.to_i)
    s = load_snapshots
    w = window(s, "seven_day", now, WEEK + 86_400)
    w&.merge(age: data_age(s, now))
  end

  def five_hour(now = Time.now.to_i)
    s = load_snapshots
    w = window(s, "five_hour", now, 6 * 3_600)
    w&.merge(age: data_age(s, now))
  end

  # LINEAR pace: allowance = the fraction of the 7-day window elapsed (not ceil'd days --
  # that front-loads and reads too rosy). delta>0 = ahead of pace (burning fast).
  def pace(used, reset, now = Time.now.to_i)
    elapsed = (now - (reset - WEEK)).clamp(0, WEEK)
    frac = elapsed.to_f / WEEK
    { elapsed_pct: frac * 100, delta: used - frac * 100, to_reset: reset - now }
  end
end
